clang/lib/Serialization/ASTReader.cpp

```diff
@@ -10057,15 +10057,18 @@ void ASTReader::finishPendingActions() {
       // For a function defined inline within a class template, force the
       // canonical definition to be the one inside the canonical definition of
       // the template. This ensures that we instantiate from a correct view
-      // of the template.
+      // of the template. This behaviour seems to be important only for inline
+      // friend functions. For normal member functions, it might results in
+      // selecting canonical decl from module A but body from module B.
       //
       // Sadly we can't do this more generally: we can't be sure that all
       // copies of an arbitrary class definition will have the same members
       // defined (eg, some member functions may not be instantiated, and some
       // special members may or may not have been implicitly defined).
-      if (auto *RD = dyn_cast<CXXRecordDecl>(FD->getLexicalParent()))
-        if (RD->isDependentContext() && !RD->isThisDeclarationADefinition())
-          continue;
+      if (FD->getFriendObjectKind())
```

Comment:
Does it mean we can still potentially have this problem for `friend` functions?
Or are there some invariants in Clang/C++ itself that stop that from happening?

If we have a small example, it'd be interesting to check that adding `friend` does not cause the problem to resurface.

Reply:
Yeah, I'm not sure that there is no sequence of deserialization that might also lead to similar issue. The whole logic about which lazy desrialization and which decl will be canonical and finally definition is very fragile I think :(

Initially I was thinking about removing the whole if because it can transform canonical decl with body may not be a definition after deserialization. But removing this logic completely breaks this [test case](https://github.com/llvm/llvm-project/blob/main/clang/test/Modules/friend-definition-2.cpp). But it seems that it was added only for friend decls and nothing else.

Reply:
The original comments confuses me. I feel it is not consistent (or at least not straight forward) to the codes. I agree the original mechanism is fragile. 

Could you try to describe the problem in more detail? Let's try to fix the fundamental problem if possible.

Reply:
@ChuanqiXu9 and @ilya-biryukov I'm not sure that I fully understand the example [test case](https://github.com/llvm/llvm-project/blob/main/clang/test/Modules/friend-definition-2.cpp). Do you understand what does the example really do?

Things unclear to me marked with `// >>>`.
```
// RUN: %clang_cc1 -std=c++14 -fmodules %s -verify
// RUN: %clang_cc1 -std=c++14 -fmodules %s -verify -triple i686-windows
// expected-no-diagnostics

// >>> it looks like undocumented pragma?
#pragma clang module build A 

// >>> Is it modulemap file syntax?
module A {}
#pragma clang module contents
#pragma clang module begin A

// >>> What is the context of this code, is it GMF or part of module A, is it implicitly exported?
template<typename T> struct ct { friend auto operator-(ct, ct) { struct X {}; return X(); } void x(); };
#pragma clang module end
#pragma clang module endbuild

#pragma clang module build B
module B {}
#pragma clang module contents
#pragma clang module begin B

// >>> Seems like it was part of GMF otherwise it is duplicating declaration?
template<typename T> struct ct { friend auto operator-(ct, ct) { struct X{}; return X(); } void x(); };
inline auto f() { return ct<float>() - ct<float>(); }
#pragma clang module end
#pragma clang module endbuild

// >>> Seems like here we are in different translation unit

// Force the definition of ct in module A to be the primary definition.
#pragma clang module import A
// >>> Why does it define something from module A !?
// >>> Should we see here error like: `declaration of 'x' in the global module follows declaration in module A`
template<typename T> void ct<T>::x() {}

// Attempt to cause the definition of operator- in the ct primary template in
// module B to be the primary definition of that function. If that happens,
// we'll be left with a class template ct that appears to not contain a
// definition of the inline friend function.
#pragma clang module import B
auto v = f();

ct<int> make();
void h() {
  make() - make();
}
```

Reply:
Yeah let's try to understand the fundamental issue and fix it fundamentally.

> // >>> it looks like undocumented pragma?

This is the pragma for clang modules. I am not familiar with it too @ilya-biryukov is it possible to get some google forks familiar with this involved?

> // >>> What is the context of this code, is it GMF or part of module A, is it implicitly exported?

It is not GMF, nor a named module in C++20 sense nor exported in C++20 sense. It is a clang module. So the concept of `GMF`, named module and export ness (and the decl level) is not meaningful here.

> // >>> Seems like it was part of GMF

It won't be a GMF for sure.

> // >>> Why does it define something from module A !?

So given it is not the named module in C++20, so the ODR rule in C++20 doesn't apply here. Correct me if I am wrong. I feel clang modules implement the header's semantics, and the name of their modules are simply an annotation and doesn't have semantical effect. So I am not surprise that it is a valid code since it basically a concentration of two headers.

Reply:
> // >>> it looks like undocumented pragma?

The pragma is coming from `-frewrite-imports`, which is a way to generate multi-module reproducers based in a single file. It does the "obvious" thing: reads the module map until `#pragma clang module contents`, then builds a unique PCM from the code up until `#pragma clang module end`, which can later be imported by `#pragma clang module import <Name>`, where `<Name>` should match with what's written in `#pragma clang module build <Name>`.

// >>> Is it modulemap file syntax?

Yes!

This testcase should probably be written to use split-file, it's clearly more common these days, so having these pragmas in tests unrelated to those pragmas seems like a bad idea.


re the Clang modules: @ChuanqiXu9's about Clang modules are correct here. It's okay to have multiple declarations in different Clang modules as long as they follow ODR (consist of the same tokens, etc). They will be "merged" together (I believe this is similar to what's happening when we import two C++20 modules and both have the same declaration in the GMF; but I'm on the contrary, not an expert in C++20 modules, so please correct me if I'm wrong).

The Clang modules are not a concatenation of two headers, rather implement "merging" of ASTs from those two headers, as if each class was defined with a unique header guard and consecutive `#include` would only skip over the second one (in fact, we even have optimizations that skip over class tokens if the same class was already imported in some Clang module).

Please let me know if that helps and if there are still unclear things in that test.

Reply:
I've sent #112380 to rewrite this test with split-file, hope that helps to get a better understanding of what's happening there.